Fix handle missing DAGs when clearing task instances to prevent failures.

jieyao-MilestoneHub opened the pull request:
This PR handle missing DAGs when clearing task instances to prevent failures. When we delete DAG files and attempt to trigger them again, the process should access the DAGs and display an error indicating that the DAGs no longer exist, rather than causing the frontend to show an "Oops" error.

### Summary of Changes:
- Added warning logs for missing DAGs to prevent silent failures during task clearing.
- Improved error handling for downstream task clearing, including checks for partial_subset errors to handle cases where DAGs or downstream dependencies are unavailable.
- Updated the docstring for _clear_task_instances for better clarity and consistent style.
- Refined the grouping and handling of Task Instances by DAG and DAG Run to improve robustness and maintainability.

### Issue Reference:

Closes https://github.com/apache/airflow/issues/44274

^ Add meaningful description above
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines) for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named {pr_number}.significant.rst or {issue_number}.significant.rst, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Commit:
Merge: new main

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
fix: handle missing DAGs and errors in `_clear_task_instances`

    - Added warning logs for missing DAGs to prevent silent failures during task clearing.
    - Improved error handling for downstream task clearing, including checks for partial_subset errors to handle cases where DAGs or downstream dependencies are unavailable.
    - Updated the docstring for _clear_task_instances for better clarity and consistent style.
    - Refined the grouping and handling of Task Instances by DAG and DAG Run to improve robustness and maintainability.

Commit:
Merge remote-tracking branch 'upstream/main'

pierrejeambrun reviewed (commented):
I don't know if the code should succeed and proceed when the dag file has been removed. Maybe a clearer error message would be more appropriate ?

I'll let other weight in, this will end up in a partial success of the request. (maybe we should fail everything, returning the tasks that cause trouble, and expect the user to remove those tasks from the future request)

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
fix: handle missing DAGs gracefully in task instance clearing

This commit addresses the issue where missing DAGs cause failures during task instance clearing. Instead of proceeding with partial success, the operation now raises a clear error and halts if any DAG is not found.

Summary of Changes:
- `_clear_task_instances` now checks for DAG existence. If a DAG is missing, an `AirflowException` is raised with a user-friendly error message.
- Updated `action_clear` and `action_clear_downstream` to catch `AirflowException` and display appropriate error messages in the UI.
- Improved logging for scenarios where downstream dependencies cannot be processed.

This change ensures that users are informed when DAGs are missing and provides clarity on why clearing task instances might fail.

Commit:
Merge branch 'main' into feature/fix-task-instance-dag-delete

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
fix: delete new tests

Commit:
Merge branch 'main' into feature/fix-task-instance-dag-delete

pierrejeambrun commented on airflow/www/views.py:
```diff
@@ -5298,42 +5298,50 @@ def _clear_task_instances(
         for dag_id, dag_tis in tis_grouped_by_dag_id:
             dag = get_airflow_app().dag_bag.get_dag(dag_id)
 
+            if not dag:
+                raise AirflowException(f"DAG with id '{dag_id}' does not exist. Cannot clear tasks.")
+
```
I don't think this will behave correctly.

Each item of the loop will proceed in success until one of the `dag` is missing and then the process will stop. 

Earlier dag will have succeeded and the transaction is committed. Then it is hard for the user to know what dag left to clear, what's not.

I would take this check first on all dags:
```
dag = get_airflow_app().dag_bag.get_dag(dag_id)
```

Accumulate all the missing dag and tasks. And in case we have any, return something like:
```
AirflowException(f"DAGs with ids {dag_ids} do not exist. Related tasks {task_ids} cannot be cleared.")
```
`dag_ids` is ids of all the missing dags. `task_ids` is the ids of all tasks related to these missing dags.

(Either everything succeed, or nothing does, the user has information to remove the tasks belonging to missing dags and proceed with the clear)

That's just an idea.

pierrejeambrun replied:
Also it would be great to have a test for that. (clear tasks from 2 dags, 1 dag is missing, not the other. Nothing should be cleared, and the returned message should state what dag and tasks caused the failure)
